### PR TITLE
Pull portability macros from feature/nvcxx_compatibility

### PR DESCRIPTION
**Pull request by wmaxey**

### CODE REVIEW

**Comment by wmaxey** [comment]
APIs provided by this:

* `NV_IF_TARGET(cond, t, ...)`
* `NV_DISPATCH_TARGET(...)`

Internally used *ONLY* to enable/disable headers and features:
`_NV_TARGET_MINIMUM_VAL` - Provides a preprocessor time integer value representing the lowest SM architecture supplied at compile time. This behaves very differently on NVCC and NVC++.

File: include/nv/detail/__target_macros [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+#  define _NV_TARGET_VAL_SM_60 nv::target::sm_60
+#  define _NV_TARGET_VAL_SM_61 nv::target::sm_61
+#  define _NV_TARGET_VAL_SM_62 nv::target::sm_62
+#  define _NV_TARGET_VAL_SM_70 nv::target::sm_70
+#  define _NV_TARGET_VAL_SM_72 nv::target::sm_72
+#  define _NV_TARGET_VAL_SM_75 nv::target::sm_75
+#  define _NV_TARGET_VAL_SM_80 nv::target::sm_80
+#  define _NV_TARGET_VAL_SM_86 nv::target::sm_86
+
+#  define _NV_TARGET_VAL nv::target::device
```
Comment (dkolsen-pgi) [inline include/nv/detail/__target_macros:31]: `nv::target::device` is no longer defined.  Which makes me think that `_NV_TARGET_VAL` is no longer used in the non-NVCC branch.  `_NV_TARGET_VAL` should be removed from the NVC++ section.

File: include/nv/detail/__target_macros [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+
+#  define _NV_TARGET___NV_IS_HOST nv::target::is_host
+#  define _NV_TARGET___NV_IS_DEVICE nv::target::is_device
+
+#  define _NV_TARGET___NV_ANY_TARGET (nv::target::is_host || nv::target::is_device)
+#  define _NV_TARGET___NV_NO_TARGET !_NV_TARGET___NV_ANY_TARGET
+
+#if defined(__NVCOMPILER_CUDA_ARCH__)
```
Comment (dkolsen-pgi) [inline include/nv/detail/__target_macros:39]: `__NVCOMPILER_CUDA_ARCH__` will go away once Allison finishes refactoring CUB.  It shouldn't be referenced at all by `<nv/target>`.  I think `_NV_TARGET_DEF` is not necessary.  `_NV_TARGET_MINIMUM_VAL` should be kept, but on its own and not within a `#if defined(__NVCOMPILER_CUDA_ARCH__)` block.

File: include/nv/target [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#if defined(__NVCC__) || defined(__CUDACC_RTC__)
+#  define _NV_COMPILER_NVCC
+#elif defined(__PGIC__)
```
Comment (dkolsen-pgi) [inline include/nv/target:14]: `#elif defined(__NVCOMPILER)`

`__PGIC__` is old name that shouldn't be used anymore.

File: include/nv/detail/__target_macros [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+#  define _NV_TARGET_VAL_SM_60 600
+#  define _NV_TARGET_VAL_SM_61 610
+#  define _NV_TARGET_VAL_SM_62 620
+#  define _NV_TARGET_VAL_SM_70 700
+#  define _NV_TARGET_VAL_SM_72 720
+#  define _NV_TARGET_VAL_SM_75 750
+#  define _NV_TARGET_VAL_SM_80 800
+#  define _NV_TARGET_VAL_SM_86 860
+
+// There is a case where __CUDA_ARCH__ is 0 while undefined that can cause host code paths to emit a device code dispatch entry
```
Comment (dkolsen-pgi) [inline include/nv/detail/__target_macros:106]: Get rid of this comment.  It doesn't apply here.

File: .upstream-tests/test/cuda/test_platform.pass.cpp [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+//
+//===----------------------------------------------------------------------===//
+
+#include <nv/target>
+#include <stdio.h>
+#include <assert.h>
+
+#if defined(__NVCC__)
+#  define TEST_NVCC
+#elif defined(__PGIC__)
```
Comment (dkolsen-pgi) [inline .upstream-tests/test/cuda/test_platform.pass.cpp:16]: `#elif defined(__NVCOMPILER)`

**Review by dkolsen-pgi — CHANGES_REQUESTED** [review]

**Comment by brycelelbach** [comment]
Can we add `nv::target::any_target` and `nv::target::no_target`.

File: include/nv/detail/__target_macros (hunk 2 above)
Comment (wmaxey) [reply include/nv/detail/__target_macros:39]: I don't see it being used in the code base at this point

**Comment by wmaxey** [comment]
I removed a couple commits that were redundant. I've implemented `nv::target::any_target` and `no_target`

@brycelelbach or @dkolsen-pgi for re-review.

File: include/nv/target [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+
+    // The predicates for basic host/device selection
+    constexpr target_description is_host =
+      target_description(detail::all_hosts);
+    constexpr target_description is_device =
+      target_description(detail::all_devices);
+    constexpr target_description any_target =
+      target_description(detail::all_hosts | detail::all_devices);
+    constexpr target_description no_target =
+      target_description(detail::all_hosts & detail::all_devices);
```
Comment (dkolsen-pgi) [inline include/nv/target:133]: This should just be `no_target = target_description(0)`

**Comment by wmaxey** [comment]
Local tests for all supported host compilers are clear. GCC, clang, and MSVC happily produce host only code.

**Review by dkolsen-pgi — APPROVED** [review]

**Pull request merged**